### PR TITLE
[graviton][pytorch][build] graviton pytorch 1.10

### PR DESCRIPTION
**Pull request by Qingzi-Lan**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron"/"graviton">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/buildspec-1-9-1-graviton.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
Comment (tejaschumbalkar) [inline pytorch/buildspec-1-9-1-graviton.yml:1]: not needed.

File: pytorch/buildspec-graviton.yml [hunk 2]
```diff
@@ -1,9 +1,9 @@
 account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
 region: &REGION <set-$REGION-in-environment>
 framework: &FRAMEWORK pytorch
-version: &VERSION 1.9.0
+version: &VERSION 1.10.0
 os_version: &OS_VERSION ubuntu20.04
-short_version: &SHORT_VERSION 1.9
+short_version: &SHORT_VERSION "1.10"
```
Comment (tejaschumbalkar) [inline pytorch/buildspec-graviton.yml:6]: nit: quotes can be removed

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+# install torch and torchvision from the PyTorch website
+RUN pip uninstall -y torch \
+ && pip install --no-cache-dir -U $PT_INFERENCE_URL
+RUN pip install torchvision
+
+## compile and install ArmCL
+RUN git clone https://github.com/Arm-software/ComputeLibrary.git \
+ && cd ComputeLibrary \
+ && git checkout v21.05 \
```
Comment (tejaschumbalkar) [inline pytorch/inference/docker/1.10/py3/Dockerfile.graviton:108]: lets parameterize the git branch. is there a reason to use the specific branch?

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+RUN HOME_DIR=/root \
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl https://aws-dlc-licenses.s3.amazonaws.com/pytorch-1.9/license.txt -o /license.txt
```
Comment (tejaschumbalkar) [inline pytorch/inference/docker/1.10/py3/Dockerfile.graviton:148]: license file should be for PT1.10

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+ && pip install --no-cache-dir -U $PT_INFERENCE_URL
+RUN pip install torchvision
+
+## compile and install ArmCL
+RUN git clone https://github.com/Arm-software/ComputeLibrary.git \
+ && cd ComputeLibrary \
+ && git checkout v21.05 \
+ && scons Werror=0 -j8 debug=0 neon=1 opencl=0 os=linux arch=arm64-v8a build=native \
+ && mkdir /opt/armClArtifact \
+ && cp build/libarm* /opt/armClArtifact
```
Comment (tejaschumbalkar) [inline pytorch/inference/docker/1.10/py3/Dockerfile.graviton:111]: on a side note, do we need to cleanup files created as a result of the armcl build?

**Comment by arjkesh** [comment]
Are we planning to tag these as e3?

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton [hunk 6]
```diff
@@ -0,0 +1,152 @@
+FROM arm64v8/ubuntu:20.04
```
Comment (tejaschumbalkar) [inline pytorch/inference/docker/1.10/py3/Dockerfile.graviton:1]: Lets remove all the SM related components from the file

File: pytorch/buildspec-graviton.yml (hunk 2 above)
Comment (Qingzi-Lan) [reply pytorch/buildspec-graviton.yml:6]: I tried 1.10 without quotes, but the path will automatically become 1.1. 
I found the quotes can solve this issue from the x86 pytorch 1.10 yml file

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton (hunk 3 above)
Comment (Qingzi-Lan) [reply pytorch/inference/docker/1.10/py3/Dockerfile.graviton:108]: for graviton pytorch 1.9 wheel, it was build on ArmCL v21.05.
library build from other branches are not compatible with the wheel.
will parameterize the branch name

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton (hunk 5 above)
Comment (Qingzi-Lan) [reply pytorch/inference/docker/1.10/py3/Dockerfile.graviton:111]: I think we should in order to minimize the size of DLC. will do it.

**Comment by Qingzi-Lan** [comment]
> Are we planning to tag these as e3?

you mean tag this PR or the DLC as e3?

**Comment by arjkesh** [comment]
> > Are we planning to tag these as e3?
> 
> you mean tag this PR or the DLC as e3?

I mean tag the DLC as e3

**Comment by Qingzi-Lan** [comment]
> > > Are we planning to tag these as e3?
> > 
> > 
> > you mean tag this PR or the DLC as e3?
> 
> I mean tag the DLC as e3

I think it can be tagged as e3, @tejaschumbalkar what do you think?

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton [hunk 7]
```diff
@@ -96,11 +95,22 @@ RUN conda install -c \
...
-RUN pip install torch torchvision
+RUN pip uninstall -y torch \
+ && pip install --no-cache-dir -U $PT_INFERENCE_URL
+RUN pip install torchvision
+
+## compile and install ArmCL
+RUN git clone https://github.com/Arm-software/ComputeLibrary.git \
+ && cd ComputeLibrary \
+ && git checkout $ARMCL_BRANCH \
+ && scons Werror=0 -j8 debug=0 neon=1 opencl=0 os=linux arch=arm64-v8a build=native \
```
Comment (tejaschumbalkar) [inline pytorch/inference/docker/1.10/py3/Dockerfile.graviton:106]: lets also remove `scons` after the ArmCL binary is generated as it adds no value to the image after that.

Comment (tejaschumbalkar) [reply pytorch/inference/docker/1.10/py3/Dockerfile.graviton:106]: > > > > Are we planning to tag these as e3?
> > > 
> > > 
> > > you mean tag this PR or the DLC as e3?
> > 
> > 
> > I mean tag the DLC as e3
> 
> I think it can be tagged as e3, @tejaschumbalkar what do you think?

lets create a separate PR for DIY/SM split.

**Review by tejaschumbalkar — APPROVED** [review]
